packages/oauth/src/tokenSession.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Juno contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { parseIdTokenData, decodeIDToken } from "./tokenHelpers"
+
+interface AuthData {
+  JWT: string
+  raw: Record<string, any>
+  refreshToken: string
+  parsed: any // Adjust this type as per the return type of parseIdTokenData
+}
+
+export const composeAuthData = (token: string, options: any): AuthData => {
+  let tokenData = {}
+  try {
+    tokenData = decodeIDToken(token)
+  } catch (_) {
+    console.warn(`WARNING: (OAUTH) Could not parse token: ${token}`)
+  }
+
+  // extend tokenData with the options
+  tokenData = { ...tokenData, ...options }
+
+  return {
+    JWT: token,
+    raw: tokenData,
+    refreshToken: "TOKEN",
+    parsed: parseIdTokenData(tokenData),
+  }
+}
+
+interface TokenSessionReturnType {
+  login: () => void
+  logout: () => void
+  currentState: () => Record<string, any>
+}
+
+export default function tokenSession(params: any = {}): TokenSessionReturnType {
+  const { token, onUpdate, options, ...unknownProps } = params || {}
+
+  if (typeof onUpdate !== "function") {
+    throw new Error("(OAUTH MOCK) onUpdate should be a function")
+  }
+
+  if (Object.keys(unknownProps).length > 0) {
```

Comment:
You could also consider creating an options guard. The zod library is excellent for this purpose. It’s just a suggestion and not a blocking change. 🙂

```ts
import { z } from "zod";

// Define the schema
const optionsSchema = z.object({
  token: z.string().optional(),
  onUpdate: z.function().optional(),
});

// Validate unknownProps
function validateOptions(options: unknown) {
  const result = optionsSchema.safeParse(options);

  if (!result.success) {
    const unknownKeys = Object.keys(options).filter(
      (key) => !(key in optionsSchema.shape)
    );

    console.warn(
      `WARNING: (OAUTH) Unknown options: ${unknownKeys.join(", ")}. Allowed options are: ${Object.keys(optionsSchema.shape).join(", ")}`
    );
  }

  return result;
}
```

Reply:
Sounds good. I didn't know about this lib, but makes sense

Reply:
fixed with types

Reply:
Type guards are meant for runtime checks, not just development. During development, TypeScript already ensures type safety, but at runtime, type guards provide additional validation to ensure your program handles data correctly in dynamic scenarios. ;)